teps/0081-add-chains-subcommand-to-the-cli.md

```diff
@@ -0,0 +1,326 @@
+  Avoid getting hung up on specific details and instead aim to get the goals of
+  the TEP clarified and merged quickly.  The best way to do this is to just
+  start with the high-level sections and fill out details incrementally in
+  subsequent PRs.
+
+Just because a TEP is merged does not mean it is complete or approved.  Any TEP
+marked as a `proposed` is a working document and subject to change.  You can
+denote sections that are under active debate as follows:
+
+```
+<<[UNRESOLVED optional short context or usernames ]>>
+Stuff that is being argued.
+<<[/UNRESOLVED]>>
+```
+
+When editing TEPS, aim for tightly-scoped, single-topic PRs to keep discussions
+focused.  If you disagree with what is already in a document, open a new PR
+with suggested changes.
+
+If there are new details that belong in the TEP, edit the TEP.  Once a
+feature has become "implemented", major changes should get new TEPs.
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/teps/NNNN-TEP-template/README.md).
+
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This proposal aims at treating `Chains` common actions as first class citizens
+by adding a new command to the Tekton CLI.
+
+## Motivation
+
+When working with `Chains`, the user experienced will be enhanced if users are
+able to
+*  use the Tekton CLI rather than a combination of other tools like `kubectl`,
+`jq`, `base64`, etc.
+* not having to memorize which exact TaskRun annotation or key to query.
+* discover available `Chains` features via the CLI and/or its documentation.
+
+### Goals
+
+Add the ability to run `tkn chains <chains_action>` fom the CLI.
+
+### Non-Goals
+
+While the idea is to use the Tekton CLI to replace a combination of shell
+commands, we may not want to implement all exotic variations and focus only on
+the most common use cases in order to avoid overloading the `chains` command.
+
+### Use Cases
+
+#### Use Case 1
+
+Extract the Chains payload from a taskrun:
+
+```bash
+$ tkn chains payload cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
```

Comment:
I always worry a bit about printing signed payloads without verifying the signature from a UX perspective. Do you think signature verification is in scope here for the tkn cli?

If not, what do you think about something like `tkn chains unsafe-payload` just to make it abundantly clear that this hasn't been verified?

Reply:
Good point!

What about we leave the name of the sub-command to `payload` but we add a `--verify` flag which would verify the signature? The reason I am proposing to keep the sub-command name is because `tkn chains unsafe-payload --verify` sounds unclear (i.e. "asking to verify something which is unsafe" 🤔 ).

Reply:
Yeah that flag issue makes sense... I still don't love the idea of the default behavior being unsafe though. I usually prefer people to have to "opt-in" to the unsafe version.

What about maybe:

```shell
tkn chains payload --skip-verify
```

To skip it, but the default would be to verify?

Reply:
Yes, it makes complete sense 👍

Reply:
Another thought: doing so would make it a requirement to add `cosign` and `in-toto` as dependencies of the CLI, like what I described in the `Notes/Caveats` section. Is that an issue?

Reply:
We're working on a lightweight, stdlib only verification package for cosign that should work here.

Reply:
>tkn chains payload --skip-verify

This makes a lot more sense to me.
Just one question, do we see any other sub-commands to be associated with the payload.
`tkn chains payload extract --skip-verify`

Reply:
As of now, no. I don't have any other use cases to do with the payload except extracting it and verifying it.

Reply:
Then it could be `payload-extract` subcommand, but I think I prefer `payload extract` 😅

Reply:
Note: there's a cosign library in progress now that should make the dependency tree much leaner. https://github.com/sigstore/cosign/issues/666